### PR TITLE
新MapReduce接口

### PR DESCRIPTION
**Pull request by KKRainbow**
- 实现Mapper、Reducer时使用接口而非函数
- Reducer的Values使用迭代函数，更易用，并且可以自定义Reducer的并行度。
- 声明Key、Value的类型后，[]byte到其他类型的转换由executor完成。
- 保留原有API

### CODE REVIEW

File: executor_new/executor_new.go [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+	"encoding/json"
+	"errors"
+	"flag"
+	"os"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
```
Comment (Lcch) [inline executor_new/executor_new.go]: import style

File: executor/executor.go [hunk 2]
```diff
@@ -37,193 +40,38 @@ var (
...
+		keyBytes := outClass.GetOutputKeyClass().ToBytes(k)
+		valueBytes := outClass.GetOutputValueClass().ToBytes(v)
+		outputKV <- &kmrpb.KV{Key: keyBytes, Value: valueBytes}
+	}
 }
 
-func (cw *ComputeWrap) BindCombiner(combiner func(v1 []byte, v2 []byte) []byte) {
-	cw.combineFunc = combiner
+type ComputeWrapClass struct {
+	mapClass    mapred.Mapper
```
Comment (Lcch) [inline executor/executor.go]: mapper & reducer is better than mapClass and reduceClass

File: mapred/mapred.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+//InputOutputClassBase input class base
+type InputOutputClassBase interface {
+	InputClassBase
+	OutputClassBase
+}
+
+//MapperReducerBase some common func
+type MapperReducerBase interface {
+	BeforeRun()
```
Comment (Lcch) [inline mapred/mapred.go]: "Init" is better

File: executor/executor.go [hunk 4]
```diff
@@ -351,25 +212,256 @@ func (cw *ComputeWrap) doReduce(interBk bucket.Bucket, reduceID int, nMap, batch
...
+			curRecord.Value = cw.combineFunc(curRecord.Value, r.Value)
+		}
+	}
+	if curRecord.Key != nil {
+		combined = append(combined, curRecord)
+	}
+	return combined
+}
+
+func (cw *ComputeWrapClass) aggregateValues(inputs <-chan *kmrpb.KV, values chan<- interface{}) {
```
Comment (Lcch) [inline executor/executor.go]: remove it

File: executor/executor.go [hunk 5]
```diff
@@ -351,25 +212,256 @@ func (cw *ComputeWrap) doReduce(interBk bucket.Bucket, reduceID int, nMap, batch
...
+		wgForReducerCaller.Add(1)
+		go reducerCaller(inputsArr[i], &wgForReducerCaller)
+	}
+
 	for r := range sorted {
-		inputs <- RecordToKV(r)
+		if ReducerConcurrentLevel == 1 {
+			inputsArr[0] <- RecordToKV(r)
+		} else {
+			k := mmh3.Hash32(r.Key[:]) % ReducerConcurrentLevel
```
Comment (Lcch) [inline executor/executor.go]: Hash32(r.Key[])

File: executor/executor.go [hunk 6]
```diff
@@ -233,14 +81,27 @@ func (cw *ComputeWrap) doMap(rr records.RecordReader, bk bucket.Bucket, mapID in
...
+	}()
 	go func() {
 		sem := util.NewSemaphore(maxNumConcurrentFlush)
 		var waitFlushWrite sync.WaitGroup
 		for in := range outputKV {
 			aggregated = append(aggregated, KVToRecord(in))
 			currentAggregatedSize += 8 + len(in.Key) + len(in.Value)
-			if currentAggregatedSize >= *flushSize*1024*1024 {
+			// if currentAggregatedSize >= *flushSize*1024*1024 {
+			if currentAggregatedSize >= 1*1024*1024 {
```
Comment (Lcch) [inline executor/executor.go]: should be about 80M

File: executor/counters.go [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package executor
```
Comment (Lcch) [inline executor/counters.go]: why do we need this?

File: mapred/mapred.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+func (tb *MapReduceBase) GetOutputValueClass() TypeClass {
+	return tb.OutputValueClass
+}
+
+//BeforeRun default empty
+func (tb *MapReduceBase) BeforeRun() {
+}
+
+//ForEachValue iterate values using iterator
+func ForEachValue(valuesNext func() (interface{}, error), handler func(interface{})) {
```
Comment (Lcch) [inline mapred/mapred.go]: move this function to exector.go

File: mapred/keyvalue.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
+package mapred
+
+import (
+	"bytes"
+	"encoding/binary"
+)
+
+//Integer64 int64
```
Comment (Lcch) [inline mapred/keyvalue.go]: add a space between // and the first character for all line comments

File: mapred/keyvalue.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+//TypeClass Interface of both key and value
+type TypeClass interface {
+	FromBytes(b []byte) interface{}
+	ToBytes(v interface{}) []byte
+}
+
+//FromBytes get value of int64, little endian
+func (v Uint64) FromBytes(b []byte) interface{} {
+	if len(b) != 8 {
+		panic(b)
```
Comment (Lcch) [inline mapred/keyvalue.go]: use log.Fatal instead

Comment (Lcch) [inline mapred/keyvalue.go]: same as all panic below

File: mapred/keyvalue.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+func (s String) ToBytes(str interface{}) []byte {
+	if res, ok := str.(string); ok {
+		return []byte(res)
+	}
+	panic(str)
+}
+
+//FromBytes return bytes slice
+func (bs Bytes) FromBytes(b []byte) interface{} {
+	return b[:]
```
Comment (Lcch) [inline mapred/keyvalue.go]: remove [:]

File: mapred/keyvalue.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
+package mapred
```
Comment (Lcch) [inline mapred/keyvalue.go]: "keyvalue.go" is kind of confused. type.go is better?

File: executor/executor.go [hunk 13]
```diff
@@ -351,25 +212,256 @@ func (cw *ComputeWrap) doReduce(interBk bucket.Bucket, reduceID int, nMap, batch
...
+					return nil, errors.New("EOINPUT")
+				}
+				if startKVPair != nil {
+					tmp := startKVPair.Value[:]
+					startKVPair = nil
+					return valueClass.FromBytes(tmp), nil
+				}
+				kvpair := <-inputs
+				if kvpair != nil {
+					if bytes.Equal(kvpair.Key[:], curKeyBytes[:]) {
```
Comment (Lcch) [inline executor/executor.go]: remove [:]

File: executor/counters.go (hunk 7 above)
Comment (KKRainbow) [inline executor/counters.go]: I want to record some statistic data such as how much k/v have been processed or produced, how much time has elapsed. Moreover, a [simple global counter](https://hadoop.apache.org/docs/r2.7.0/api/org/apache/hadoop/mapred/Counters.html)  can be implemented based on it.

Comment (KKRainbow) [inline executor/counters.go]: I think the executor should regard map and reduce as separated step and executor class structure will be changed. I planned to implement counter after the executor changed.

File: mapred/mapred.go (hunk 8 above)
Comment (KKRainbow) [inline mapred/mapred.go]: I think as a helper function, it does't belong to executor in logic sense.

File: executor/executor.go (hunk 2 above)
Comment (KKRainbow) [inline executor/executor.go]: Done

File: mapred/mapred.go (hunk 3 above)
Comment (KKRainbow) [inline mapred/mapred.go]: Done

File: executor/executor.go (hunk 4 above)
Comment (KKRainbow) [inline executor/executor.go]: Done

File: executor/executor.go (hunk 5 above)
Comment (KKRainbow) [inline executor/executor.go]: Done

File: executor/executor.go (hunk 6 above)
Comment (KKRainbow) [inline executor/executor.go]: Done. I use original flushSize

File: executor/executor.go (hunk 13 above)
Comment (KKRainbow) [inline executor/executor.go]: Done

File: mapred/keyvalue.go (hunk 10 above)
Comment (KKRainbow) [inline mapred/keyvalue.go]: Done, but I have to use an redundant return to please my IDE or it'll complain no return at the end of function

File: mapred/keyvalue.go (hunk 11 above)
Comment (KKRainbow) [inline mapred/keyvalue.go]: Done

File: mapred/keyvalue.go (hunk 12 above)
Comment (KKRainbow) [inline mapred/keyvalue.go]: Done

File: examples/wordcount_new/example.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+	"fmt"
+	"strings"
+	"unicode"
+
+	"github.com/naturali/kmr/executor"
+	"github.com/naturali/kmr/mapred"
+)
+
+const (
+	MAX_WORD_LENGTH = 20
```
Comment (xiang90) [inline examples/wordcount_new/example.go]: MaxWordLength. 

doc string on public const.

File: examples/wordcount_new/example.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+type WordCountMap struct {
```
Comment (xiang90) [inline examples/wordcount_new/example.go]: doc string

File: examples/wordcount_new/example.go [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+	mapred.MapReduceBase
+}
+
+type WordCountReduce struct {
+	mapred.MapReduceBase
+}
+
+func (*WordCountMap) Map(key interface{}, value interface{}, output func(k, v interface{}), reporter interface{}) {
```
Comment (xiang90) [inline examples/wordcount_new/example.go]: doc string

File: examples/wordcount_new/example.go [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+	}
+}
+
+func (*WordCountReduce) Reduce(key interface{}, valuesNext func() (interface{}, error), output func(k interface{}, v interface{}), reporter interface{}) {
+	var count uint64
+	mapred.ForEachValue(valuesNext, func(value interface{}) {
+		val, _ := value.(uint64)
+		count += val
+	})
+	fmt.Println(key, " ", count)
```
Comment (xiang90) [inline examples/wordcount_new/example.go]: remove random debug logging

File: mapred/mapred.go [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// InputOutputTypeConverter input class base
+type InputOutputTypeConverter interface {
+	InputTypeConverter
+	OutputTypeConverter
+}
+
+// MapperReducerBase some common func
+type MapperReducerBase interface {
```
Comment (xiang90) [inline mapred/mapred.go]: why? this interface is confusing.

File: mapred/mapred.go [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// Reducer Reducer interface
+type Reducer interface {
+	InputOutputTypeConverter
+	MapperReducerBase
+	Reduce(key interface{}, valuesNext func() (interface{}, error), output func(k interface{}, v interface{}), reporter interface{})
+}
+
+// MapReduceBase Mapper and Reducer base class
```
Comment (xiang90) [inline mapred/mapred.go]: there is no class in go.

File: mapred/mapred.go [hunk 20]
```diff
@@ -0,0 +1,77 @@
+package mapred
```
Comment (xiang90) [inline mapred/mapred.go]: this set of interfaces are pretty confusing. probably need some rework.

what are you trying to achieve with these interfaces?

**Pull request merged**